### ISSUE TITLE
ViewRendered2 - add "code only" option

### ISSUE THREAD

**Issue by tbpassin** [leo-editor/leo-editor#756]
**Summary**

Tom Passin: I have modified my copy of viewrendered2.py so that it can display and export just the code. I'd really like to get these changes into the code base so I won't have to re-create them with every update to Leo.  I think there must be other people who would find this capability useful.

Terry: I hope to eventually orphan all the view rendered variants with the leo-edit-pane project, but probably best not to hold your breath :-}

**Advantages of VR2** 

ViewRendered2 is very close to a good solution for literate programming. It:

1. It allows for annotation and commentary far beyond what can be included with code comments.
2. It can execute an entire tree of Python code,
3. It can export its views to html.
4. You can change from @language python to @language rest in any node (but only once per node), which is what lets you annotate code so nicely.
5. It has many of the capabilities of Jupyter notebooks. More could be added.

The viewrendered plugin can do neither 2 nor 3.

**Ways VR2 could be improved**

If you build a program tree, you can't easily export just the code so someone else can use it, or so you can run it from the command line with various options.  True, you could put the program under an @clean node, but then you couldn't have the annotations etc.  You could do it with an @file tree, I think, but still, that would be more clumsy than needed.

**New version of VR2**

I have modified my copy of viewrendered2.py so that it can display and export just the code.  After exporting it to a browser, you can just copy the page in the browser and save the entire program to a file.  This enhancement adds a new menu item to the "Options" menu.  It also omits headlines, `@` directive lines, and section lines (lines that start with "<<", like "<<imports>>".  This lets it emit code-only for `@file` trees without those non-code lines appearing.

**Comment by edreamleo** [comment]
On Thu, Mar 1, 2018 at 6:27 PM, tbpassin <notifications@github.com> wrote:

I have modified my copy of viewrendered2.py
> ​...​
> so that it can display and export just the code
> ​...​
>
​​

> I'd really like to get these changes into the code base so I won't have to
> re-create them with every update to Leo.


​Send me viewrendered2.py and I'll create a branch for your work.

Edward

**Comment by tbpassin** [comment]
Thank you, Ed!

I have renamed my version to "viewrendered2a" just to prevent any 
accidental mixup with the original.

I suppose that the main line of development is going to be on 
viewrendered, isn't it?  Especially for Qt5 compatibility? Will VR2 
eventually become orphaned because of this?

Tom Passin

**Comment by tbnorth** [comment]
On Fri, 02 Mar 2018 08:00:40 -0800
tbpassin <notifications@github.com> wrote:

> Thank you, Ed!
> 
> I have renamed my version to "viewrendered2a" just to prevent any 
> accidental mixup with the original.
> 
> I suppose that the main line of development is going to be on 
> viewrendered, isn't it?  Especially for Qt5 compatibility? Will VR2 
> eventually become orphaned because of this?

I hope to eventually 'orphan' all the view rendered variants with the
leo-edit-pane project, but probably best not to hold your breath :-}

Cheers -Terry

> Tom Passin
> 
> On 3/2/2018 3:37 AM, Edward K. Ream wrote:
> > On Thu, Mar 1, 2018 at 6:27 PM, tbpassin <notifications@github.com>
> > wrote:
> >
> > I have modified my copy of viewrendered2.py  
> >  > ​...​
> >  > so that it can display and export just the code
> >  > ​...​
> >  >  
> 
> 
>

**Comment by tbpassin** [comment]
I'm uploading my modified copy of viewrendered2.py.  I have now added code so that the code-only rendering works for Markdown as well as for ReST.  The Options menu has a new item, "Code Only".  When checked, the "Visible Code" option gets turned on, because code-only without visible code has nothing to show.

I added a setting to change the background color of the rendering panel for Markdown, because MD processing doesn't include a stylesheet.  I needed this because I've started to use a dark theme, and the panel was unreadable.  It turned out to be a bit of a kludge to get this working, but I don't see any other way  right now.

Checked only on Python 2.7.

[viewrendered2.zip](https://github.com/leo-editor/leo-editor/files/1779377/viewrendered2.zip)
